src/darwin/Framework/CHIPTests/MTRPerControllerStorageTests.m

```diff
@@ -2172,10 +2172,944 @@ - (void)doTestSubscriptionPoolWithSize:(NSInteger)subscriptionPoolSize
+        // Also _persistedClusters and _persistedClusterData has the cluster paths for all clusters in the cluster index for endpoints 0 and 2.
+        // For endpoint 1, we should only have 1 entry in the clusterIndex for descriptor cluster. Everything else must be deleted.
+        dispatch_sync(self->_storageQueue, ^{
+            XCTAssertTrue([[controller.controllerDataStore _fetchEndpointIndexForNodeID:deviceID] isEqualToArray:testEndpoints]);
+            for (NSNumber * endpoint in testEndpoints) {
+                XCTAssertNotNil(initialClusterIndex);
+                for (NSNumber * cluster in [initialClusterIndex objectAtIndex:endpoint.unsignedShortValue]) {
+                    MTRClusterPath * path = [MTRClusterPath clusterPathWithEndpointID:endpoint clusterID:cluster];
+                    if (![endpoint isEqualToNumber:testEndpoint]) {
+                        XCTAssertNotNil([controller.controllerDataStore _fetchClusterDataForNodeID:deviceID endpointID:endpoint clusterID:cluster]);
+                        XCTAssertTrue([device _persistedClusterContains:path]);
+                        XCTAssertNotNil([device _getPersistedClusterDataForPath:path]);
+                    } else {
+                        if ([cluster isEqualToNumber:notToBeDeletedClusterID]) {
+                            XCTAssertNotNil([controller.controllerDataStore _fetchClusterDataForNodeID:deviceID endpointID:endpoint clusterID:cluster]);
+                            XCTAssertTrue([device _persistedClusterContains:path]);
+                            XCTAssertNotNil([device _getPersistedClusterDataForPath:path]);
+                        } else {
+                            XCTAssertNil([controller.controllerDataStore _fetchClusterDataForNodeID:deviceID endpointID:endpoint clusterID:cluster]);
+                            XCTAssertFalse([device _persistedClusterContains:path]);
+                            XCTAssertNil([device _getPersistedClusterDataForPath:path]);
+                        }
+                    }
+                }
+            }
+        });
+
+        [reportEndExpectation fulfill];
+    };
+
+    [device unitTestInjectAttributeReport:attributeReport fromSubscription:YES];
+
+    [self waitForExpectations:@[ attributeDataReceivedExpectation, reportEndExpectation ] timeout:60];
+
+    [controller.controllerDataStore clearAllStoredClusterData];
+    NSDictionary * storedClusterDataAfterClear = [controller.controllerDataStore getStoredClusterDataForNodeID:deviceID];
+    XCTAssertEqual(storedClusterDataAfterClear.count, 0);
+
+    [controller removeDevice:device];
+    // Reset our commissionee.
+    __auto_type * baseDevice = [MTRBaseDevice deviceWithNodeID:deviceID controller:controller];
+    ResetCommissionee(baseDevice, queue, self, kTimeoutInSeconds);
+
+    [controller shutdown];
+    XCTAssertFalse([controller isRunning]);
+}
+
+- (void)test017_TestDataStoreMTRDevicePruneOrphanedAttributes
```

Comment:
See various comments above about the endpoints and clusters.

Reply:
fixed.